src/chain.h

```diff
@@ -166,6 +167,15 @@ class CBlockIndex
     //! Will be boost::none if nChainTx is zero.
     boost::optional<CAmount> nChainSproutValue;
 
+    //! Change in value held by the Sapling circuit over this block.
+    //! Not a boost::optional because this was added before Sapling activated, so we can
+    //! rely on the invariant that every block before this was added had nSaplingValue = 0.
+    CAmount nSaplingValue;
```

Comment:
sprout pre-overwinter blocks have `nSaplingValue` defined?

Reply:
Yes, to make the

Reply:
Yes; it's a trivial result that blocks created before Sapling activates have zero effect on the Sapling value pool. It makes the semantics simpler, and means we only add 8 bytes per block in disk serialization instead of 9.

Reply:
What is the difference from `nChainSaplingValue`?  That you *do* make a boost::optional?

Reply:
`nChainSaplingValue` is the sum from genesis of the `nSaplingValue` deltas, so we can only set it on a block if we know every one of its parents. `nChainTx` is the analogous value we inherited from Bitcoin; in that case it doesn't need a `boost::optional` because there will never have been zero transactions on the chain (genesis block contains one). For us, we *will* have `nChainSaplingValue = 0` before activation, so for unconnected blocks we need the additional `boost::none` state.

Reply:
Just for sake of argument, not sure why it's more of a problem to define `nChainSaplingValue` as 0 for an unconnected block than it is for `nChainTx`,
which is what seems to happen here https://github.com/zcash/zcash/pull/3255/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR3198

Reply:
Ahh maybe that's not an unconnected block there,  maybe only the genesis block doesn't have `pprev` defined?

Reply:
@arielgabizon ``nChainSaplingValue`` can be zero for a connected block (it is for all blocks before the Sapling activation, and possibly for a little after), so that would be ambiguous. There's a case for also using ``boost::none`` rather than ``0`` for an unknown ``nChainTx`` value, but that would be a separate refactoring if we wanted to do it.

Reply:
Ahh, right it would be ambiguous to use 0 for `none`, I guess that's what @str4d.